### ISSUE TITLE
Questions about UaaTokenEnhancer

### ISSUE THREAD

**Issue by jhamon** [cloudfoundry/uaa#887]
Hey @strehle @tack-sap @torsten-sap we recently ran into some regressions related to UAA token enhancer and am wondering if you can clarify for us some things about how it is used. The tests that were contributed along with the feature aren't enough for me to understand how this is meant to be used. Are you expecting to modify claims in the token? Or only add additional claims?

The way this is currently implemented relies on [copying maps of data](https://github.com/cloudfoundry/uaa/blob/4014f6ab89b053733b77de4c258f6d61cceacb96/server/src/main/java/org/cloudfoundry/identity/uaa/oauth/UaaTokenServices.java#L273-L278) and later [naively merging them](https://github.com/cloudfoundry/uaa/blob/4014f6ab89b053733b77de4c258f6d61cceacb96/server/src/main/java/org/cloudfoundry/identity/uaa/oauth/UaaTokenServices.java#L514) with the access token claims in UaaTokenServices.  We're finding this approach hard to understand and hard to maintain. For example, if you want to quickly answer the question of what claims we should expect to appear in the JWT access tokens issued by the UAA, it is not easy to get a clear answer. If you want to make a change in refresh tokens but not access tokens, it's easy to accidentally have it appear in access tokens later where it is not wanted. This approach is error prone and has led to some unintended regressions when attributes that should not propagate forward via the refresh flow are copied (e.g. the access token `scope` and `jti` claims from the refresh token should not propagate through now that we've identified security implications of these values being copied).

When I look at the [UaaTokenEnhancer interface](https://github.com/cloudfoundry/uaa/blob/e3581e8434f41d03ce9710da2686b8fe6ee35112/server/src/main/java/org/cloudfoundry/identity/uaa/oauth/UaaTokenEnhancer.java#L16), I see the `enhance` base implementation is referencing the `ext_attr` claim. Since SAP are the only folks using this to my knowledge, I wanted to understand whether you in your TokenEnhancer implementation are also setting this `ext_attr` claim? Are you modifying any "standard" claims like `scope`, `jti`, `grant_type`, etc? 

I'd like to think about how we could change this interface to clarify the contract between the enhancer and the resultant tokens so we can prevent unintentionally leaking/copying of data as our implementation of refresh tokens and access tokens evolves.

cc: @tnwang @jspawar

**Comment by strehle** [comment]
Hi @jhamon ,

the interface of token enhance was extended with enhance and this method replaces former (existing) method getExternalAttributes.
We did the enhancements with help of @fhanik  and he pointed out that we should call token enhancer only once and use this result to enhance access and refresh tokens.
Thats why we requested https://github.com/cloudfoundry/uaa/pull/768

No, we dont modify existing claims, but extend JWT only, therefore the name enhancer. However, in discussions with @fhanik he pointed out, that with this program API we can also modify or delete claims.
But if you see where the call is done, 
https://github.com/cloudfoundry/uaa/blob/develop/server/src/main/java/org/cloudfoundry/identity/uaa/oauth/UaaTokenServices.java#L532-L534
then you see that we could modify JTI but not other entries.

We wanted a new method because with the interface of https://github.com/cloudfoundry/uaa/blob/develop/server/src/main/java/org/cloudfoundry/identity/uaa/oauth/UaaTokenEnhancer.java#L12 we cannot add arrays and we add SAML attributes with new method.

What is missing from our point now is that enhancement does not work for id_token and with K8s we need (want) use id_tokens.

Hope this helps.

regards,
-markus
